### ISSUE TITLE
resolved: MulticastDNS hostname resolution through (tinc) VPN does not work

### ISSUE THREAD

**Issue by florianjacob** [systemd/systemd#6313]
### Submission type

 - Bug report

### systemd version the issue has been seen with

> 233

### Used distribution

> Arch Linux and NixOS

### In case of bug report: Expected behaviour you didn't see

> After adding `MulticastDNS = true` to the network file which configures the vpn interface, I expected MulticastDNS hostname resolution to work from both sides, i.e. `ping <host on vpn>.local` should resolve other hosts on the VPN as well as for other hosts on the network, the local host should be resolvable. (Both works with avahi in the same VPN, after enabling the `allow-point-to-point=yes` option.)

### In case of bug report: Unexpected behaviour you saw

> Resolving times out. According to wireshark, no QUERY messages are sent from the local host on the VPN interface, only on the ethernet interface. QUERY messages from other systems on the VPN are received, but remain unanswered. It seems like resolved is ignoring the interface for MulticastDNS purposes.

### In case of bug report: Steps to reproduce the problem

> 1. Have a (tinc?) VPN set up between the local host using resolved and any other host having resolved or avahi running for MulticastDNS
> 2. Add `MulticastDNS = true` to the network file which configures the VPN, reload networkd and resolved
> 3. Try to ping the other host from either hosts

### Additional information:
```
$ ip link
…
4: tinc.wolkenplnt: <POINTOPOINT,MULTICAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP mode DEFAULT group default qlen 500
link/none

$ busctl introspect org.freedesktop.resolve1 /org/freedesktop/resolve1/link/_3<the link index>
NAME                                TYPE      SIGNATURE RESULT/VALUE      FLAGS
org.freedesktop.DBus.Introspectable interface -         -                 -
.Introspect                         method    -         s                 -
org.freedesktop.DBus.Peer           interface -         -                 -
.GetMachineId                       method    -         s                 -
.Ping                               method    -         -                 -
org.freedesktop.DBus.Properties     interface -         -                 -
.Get                                method    ss        v                 -
.GetAll                             method    s         a{sv}             -
.Set                                method    ssv       -                 -
.PropertiesChanged                  signal    sa{sv}as  -                 -
org.freedesktop.resolve1.Link       interface -         -                 -
.Revert                             method    -         -                 -
.SetDNS                             method    a(iay)    -                 -
.SetDNSSEC                          method    s         -                 -
.SetDNSSECNegativeTrustAnchors      method    as        -                 -
.SetDomains                         method    a(sb)     -                 -
.SetLLMNR                           method    s         -                 -
.SetMulticastDNS                    method    s         -                 -
.DNS                                property  a(iay)    0                 -
.DNSSEC                             property  s         "allow-downgrade" -
.DNSSECNegativeTrustAnchors         property  as        0                 -
.DNSSECSupported                    property  b         true              -
.Domains                            property  a(sb)     0                 -
.LLMNR                              property  s         "yes"             -
.MulticastDNS                       property  s         "yes"             -
.ScopesMask                         property  t         0                 -

**Comment by poettering** [comment]
What does "systemd-resolve --status" say about the interface in question?

**Comment by florianjacob** [comment]
```
$ systemd-resolve --status
[…]
Link 4 (tinc.wolkenplnt)
      Current Scopes: none
       LLMNR setting: yes
MulticastDNS setting: yes
      DNSSEC setting: allow-downgrade
    DNSSEC supported: yes
[…]
```

**Comment by poettering** [comment]
ah, eh, so we currently explicitly exclude POINTOPOINT links from LLMNR/mDNS:

https://github.com/systemd/systemd/blob/master/src/resolve/resolved-link.c#L578

And you are using a POINTOPOINT link... 

Hmm, I wonder if it would be OK to remove this check. Any chance you can drop that one check, and see if it just makes things work for you? If so, we can drop it upstream too.

Normally PPP links are not considered to be a "local link" really, that's why it is excluded here...

**Comment by florianjacob** [comment]
> Any chance you can drop that one check, and see if it just makes things work for you? If so, we can drop it upstream too.

I compiled systemd without that check and yes, it makes things just work for me, yay! 😃 

> ah, eh, so we currently explicitly exclude POINTOPOINT links from LLMNR/mDNS

Damn, I searched for a check like this, but never noticed the interface option is called POINTOPOINT and *not* POIN**T**TOPOINT…

> Normally PPP links are not considered to be a "local link" really, that's why it is excluded here...

Normally for a good reason, I admit. Some more details on my use case as justification for allowing to turn it on in some cases:

What I'm doing here is to span an overlay network between all my machines and servers, so regardless of were I take e.g. my laptop, I can reach all of them as if they were on my local network. The latency is low enough so that MulticastDNS just works and frees me from managing `/etc/hosts` by hand. It has the additional benefit of automatically going through the overlay network when I'm remote but still using the direct connection when I'm in the same network, as `<hostname>.local` will resolve to the address with lowest latency.
